### PR TITLE
decompose encoders into serializable sensors

### LINKED ISSUE #19: separated and serializable encoders
Encoders are the only things in Comportex HTM values that are not serializable. We should fix that so that models can be saved or sent over the wire.

Currently they are created using `(reify)` which makes a closure. But the main problem is how they get their particular data inputs out of the general, amorphous input-value provided to `htm-step`. That is using `pre-transform` which applies some arbitrary function to do the extraction.

#### Proposal:

Shift the task of formatting the data for encoding outside to whatever is creating the input values. An input-value as provided to `htm-step` would then be a structured value directly providing the data required by each input's encoder. Each input already has a keyword id specified in `core/RegionNetwork` so these keys could specify the input data for each.

Example. Suppose there are 2 inputs (feeding into one or more regions), called `:main-input` and `:motor`. The former is a concatenation of a category and a coordinate encoding. The latter is a single linear number encoding.

The input-value might then look like:
```clojure
{:motor 42
 :main-input [:red {:coord [5.0 10.0], :radius 2.5}]
}
```

Each input encoder will be passed just its sub value.

The various encoder types can be made into Records or Types.

### PR DESCRIPTION
**Pull request by floybix**
Fixes #19 

Sensors are defined as `[selector encoder]`. Both selectors and encoders are serializable; so HTM models should now be fully serializable.

This requires changes to all user code to define sensors. The argument list of `regions-in-series` has changed.

Is it too confusing to have "sensors" as well as "senses" (the stateful output of sensors)?

What about htm-export ... thinking we might need a general export protocol to clear out any caches (e.g. encoder caches) and compress the model - e.g. removing the reverse index in SynapseGraphs which can be rebuilt ("import" protocol?)...

### CODE REVIEW

File: test/org/nfrac/comportex/encoders_test.cljc [hunk 1]
```diff
@@ -9,18 +9,18 @@
 
 
 (deftest encode-decode-test
-  (let [e (enc/category-encoder 10 [:a :b :c])]
-    (is (= :a (:value (first (p/decode e (frequencies (p/encode e :a)) 1)))))
+  (let [e (enc/category-encoder [10] [:a :b :c])]
+    (is (= :a (:value (first (p/dcode e (frequencies (p/encode e :a)) 1)))))
```
Comment (mrcslws) [inline test/org/nfrac/comportex/encoders_test.cljc]: Typo: "dcode"

**Comment by floybix** [comment]
D'oh. fixed.

Forgot to emphasise that I changed usage of all encoders. Probably should think more about their usability / consistency.

**Comment by mrcslws** [comment]
Nice touch, allowing keywords and vectors as selectors via extend-protocol.

The sensor + sense distinction makes sense. Speaking for myself, I am not confused. 

My gut reaction is we should remove `htm-export` now that the HTM is serializable. If at some later time there's a "our serialized models are too big" problem, we could address it at that point. The notebook no longer uses `htm-export`. I'm not super-opinionated though, I could be swayed.

One part I'm a little bit unsure of. This deals with your previous change. The PHTM methods generally return a transformed RegionNetwork. Except for `sense`, which is a way of applying a sensor to an input, and returning the bits. Did you consider instead having `sense` change the model's input bits, and `htm-activate` do a column activation based on the current input bits? Then `sense` would contain all calls to `sense-activate`, and `htm-activate` would contain all calls to `region-activate`. `htm-step` would consist of calling each of these methods in order. I might be wrong, I just now stumbled on this idea after I started this paragraph.

Just musing. This change looks good. :+1:

**Comment by floybix** [comment]
You are right, as usual. That ends up being much cleaner.

I really appreciate your help with design - it is hard.

Removed htm-export.

**Pull request merged**